numba/types/function.py

```diff
@@ -0,0 +1,296 @@
+        if obj.__name__ == '<lambda>':
+            # lambda functions cannot carry annotations, hence:
+            raise ValueError('constructing numba type instance from '
+                             'a lambda function is not supported')
+        sig = inspect.signature(obj)
+        rtype = _annotation_to_numba_type(sig.return_annotation, sig)
+        atypes = []
+        for name, param in sig.parameters.items():
+            if param.kind not in [inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                                  inspect.Parameter.POSITIONAL_ONLY]:
+                raise ValueError(
+                    'callable argument kind must be positional,'
+                    ' `%s` has kind %s' % (param, param.kind))
+            atype = _annotation_to_numba_type(param.annotation, sig)
+            atypes.append(atype)
+        ftype = FunctionProtoType(rtype, atypes)
+        return FunctionType(ftype)
+
+    raise NotImplementedError(
+        'constructing numba type from %s instance' % (type(obj)))
+
+
+def _annotation_to_numba_type(annot, sig):
+    if annot == sig.empty:
+        return numbatype(None)
+    return numbatype(annot)
+
+
+def _findparen(s):
+    """Find the index of left parenthesis that matches with the one at the
+    end of a string.
+
+    Used internally. Copied from rbc/typesystem.py.
+    """
+    j = s.find(')')
+    assert j >= 0, repr((j, s))
+    if j == len(s) - 1:
+        i = s.find('(')
+        if i < 0:
+            raise NumbaTypeParseError('failed to find lparen index in `%s`' % s)
+        return i
+    i = s.rfind('(', 0, j)
+    if i < 0:
+        raise NumbaTypeParseError('failed to find lparen index in `%s`' % s)
+    t = s[:i] + '_' * (j - i + 1) + s[j + 1:]
+    assert len(t) == len(s), repr((t, s))
+    return _findparen(t)
+
+
+def _commasplit(s):
+    """Split a comma-separated items taking into account parenthesis.
+
+    Used internally. Copied from rbc/typesystem.py.
+    """
+    lst = s.split(',')
+    ac = ''
+    p1, p2 = 0, 0
+    rlst = []
+    for i in lst:
+        p1 += i.count('(') - i.count(')')
+        p2 += i.count('{') - i.count('}')
+        if p1 == p2 == 0:
+            rlst.append((ac + ',' + i if ac else i).strip())
+            ac = ''
+        else:
+            ac = ac + ',' + i if ac else i
+    if p1 == p2 == 0:
+        return rlst
+    raise NumbaTypeParseError('failed to comma-split `%s`' % s)
+
+
+class WrapperAddressProtocol(object):
+    """Base class for Wrapper Address Protocol.
+
+    Objects that type is derived from WrapperAddressProtocol can be
+    passed as arguments to numba njitted functions where it can be
+    used as first-class functions.  As minimum, the derived types must
+    implement the two methods __wrapper_address__ and signature.
+    """
+
+    def __wrapper_address__(self, sig):
+        """Return the address of a library function with given signature.
+
+        Parameters
+        ----------
+        sig : str
```

Comment:
The call to `__wrapper_address__` is constructed from `unbox_function_type` function and to support non-string `sig` would mean that the `Type` instances would be needed to lowered as well.
Using a string to represent a `Type` value at the LLVM IR level is much easier and sufficient (as long as this is needed only for passing `Type` values, not for manipulating the values within LLVM IR level).

Reply:
I wonder if the approach using `pyapi.serialize_object` could avoid the need to lower `Type` instances?
It would be equivalent to `sig` being string except that the `__wrapper_address__` caller unserializes the pickled data to `FunctionType` and passing it to `__wrapper_address__`.

Reply:
I confirm, using `pyapi.serialize_object` allows `sig` to be `Signature`.

Reply:
Fixed, the signature must now be `Signature` instance.